esapi/esapi_util/esys_mu.c

```diff
@@ -0,0 +1,1200 @@
+
+/**
+ * Unmarshal an array of BYTE structures from a byte buffer.
+ * @param[in,out] buffer Buffer to read data from.
+ * @param[in] size Size of the buffer.
+ * @param[in,out] offset Offset inside the buffer
+ *                (being updated during marshalling).
+ * @param[in] count Number of structures to be unmarshalled.
+ * @param[out] out Structures to store the result in.
+ * @retval TSS2_RC_SUCCESS on success.
+ * @retval TSS2_ESYS_RC_BAD_REFERENCE if buffer==NULL.
+ * @retval TSS2_ESYS_RC_INSUFFICIENT_BUFFER if remaining buffer is insufficient.
+ */
+TSS2_RC
+Tss2_MU_BYTE_array_Unmarshal(
+    const uint8_t *buffer,
+    size_t size,
+    size_t *offset,
+    size_t count,
+    BYTE *dst)
+{
+    LOG_TRACE("called: count=%zu buffer=%p size=%zu offset=%p dst=%p",
+        count, buffer, size, offset, dst);
+    if (buffer == NULL) {
+        LOG_ERROR("buffer=NULL");
+        return TSS2_ESYS_RC_BAD_REFERENCE;
+    }
+
+    size_t offset_loc = (offset != NULL)? *offset : 0;
+
+    if (count > size || size - count > offset_loc) {
+        LOG_ERROR("not enough space in target buffer");
+        return TSS2_ESYS_RC_INSUFFICIENT_BUFFER;
+    }
+
+    if (dst != NULL)
+        memcpy(dst, &buffer[offset_loc], count);
+    offset_loc += count;
+
+    if (offset != NULL)
+        *offset = offset_loc;
+    return TSS2_RC_SUCCESS;
+}
+
+/**
+ * Marshal a constant of type IESYSC_RESOURCE_TYPE_CONSTANT into a byte buffer.
+ * @param[in] src constant to be marshalled.
+ * @param[in,out] buffer Buffer to write result into (may be NULL)
+ * @param[in] size Size of the buffer.
+ * @param[in,out] offset Offset inside the buffer (may be NULL.
+ * @retval TSS2_RC_SUCCESS on success.
+ * @retval TSS2_ESYS_RC_BAD_REFERENCE if src==NULL.
+ * @retval TSS2_ESYS_RC_INSUFFICIENT_BUFFER if remaining buffer is insufficient.
+ */
+TSS2_RC
+Tss2_MU_IESYSC_RESOURCE_TYPE_CONSTANT_Marshal(
+    const IESYSC_RESOURCE_TYPE_CONSTANT src,
+    uint8_t *buffer,
+    size_t size,
+    size_t *offset)
+{
+    LOG_TRACE("called: src=%"PRIx32 " buffer=%p size=%zu offset=%p", src,
+        buffer, size, offset);
+    return Tss2_MU_UINT32_Marshal(src, buffer, size, offset);
+}
+
+/**
+ * Unmarshal a constant of type IESYSC_RESOURCE_TYPE_CONSTANT from a byte buffer.
+ * @param[in,out] buffer Buffer to read data from.
+ * @param[in] size Size of the buffer.
+ * @param[in,out] offset Offset inside the buffer
+ *                (being updated during marshalling).
+ * @param[out] dst variable to store the result in.
+ * @retval TSS2_RC_SUCCESS on success.
+ */
+TSS2_RC
+Tss2_MU_IESYSC_RESOURCE_TYPE_CONSTANT_Unmarshal(
+    const uint8_t *buffer,
+    size_t size,
+    size_t *offset,
+    IESYSC_RESOURCE_TYPE_CONSTANT *dst)
+{
+    LOG_TRACE("called: buffer=%p size=%zu offset=%p dst=%p",
+        buffer, size, offset, dst);
+    size_t offset_loc;
+    IESYSC_RESOURCE_TYPE_CONSTANT dst_loc;
+    TSS2_RC ret = Tss2_MU_UINT32_Unmarshal(buffer, size,
+        &offset_loc, &dst_loc);
```

Comment:
`offset_loc` passed uninitialized here. If it's not '0' then things will get weird here I think.